### PR TITLE
Dual Stack networking support

### PR DESCRIPTION
**Pull request by scholzj**
### Type of change

- Enhancement / new feature

### Description

This PR adds better support for dual stack networking (IPv4+IPv6). It lets users to choose the preferred policy and the preferred stack. This has to be done on two places:
* All internal Kafka listeners (both `type: internal` but also replication and control plane listeners share the `...-kafka-bootstrap` and `...-kafka-brokers` services), as well as Zookeeper, Connect, CruiseControl etc., have this configuration in the `template` section.
* External Kafka listener have their own services on a per listener basis and have this configuration in the `listeners` section to allow different configuration for each listener.

This PR also adds support for Enum based Lists to our CRDs and updates the `ServiceOperator` to handle the dual stack networking defaults in case user doesn't configure anything (since in that case, Kubernetes will pick some defaults on their own).

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Update CHANGELOG.md

### CODE REVIEW

File: api/src/main/java/io/strimzi/api/kafka/model/template/InternalServiceTemplate.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import io.sundr.builder.annotations.Buildable;
+import lombok.EqualsAndHashCode;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Representation of a template for Strimzi external services.
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/template/InternalServiceTemplate.java:22]: Did you mean internal services here?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java [hunk 2]
```diff
@@ -798,7 +790,8 @@ public Service generateService() {
                     ports,
                     getLabelsWithStrimziName(name, Util.mergeLabelsOrAnnotations(templateExternalBootstrapServiceLabels, ListenersUtils.bootstrapLabels(listener))),
                     getSelectorLabels(),
-                    Util.mergeLabelsOrAnnotations(ListenersUtils.bootstrapAnnotations(listener), templateExternalBootstrapServiceAnnotations)
+                    Util.mergeLabelsOrAnnotations(ListenersUtils.bootstrapAnnotations(listener), templateExternalBootstrapServiceAnnotations),
+                    ListenersUtils.ipFamilyPolicy(listener), ListenersUtils.ipFamilies(listener)
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java:794]: Can we put the args of different lines?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java [hunk 3]
```diff
@@ -864,7 +857,8 @@ public Service generateService() {
                     ports,
                     getLabelsWithStrimziName(name, Util.mergeLabelsOrAnnotations(templatePerPodServiceLabels, ListenersUtils.brokerLabels(listener, pod))),
                     selector,
-                    Util.mergeLabelsOrAnnotations(ListenersUtils.brokerAnnotations(listener, pod), templatePerPodServiceAnnotations)
+                    Util.mergeLabelsOrAnnotations(ListenersUtils.brokerAnnotations(listener, pod), templatePerPodServiceAnnotations),
+                    ListenersUtils.ipFamilyPolicy(listener), ListenersUtils.ipFamilies(listener)
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java:861]: Can we put the args of different lines?

**Review by tombentley — COMMENTED** [review]
Could we add a test for the list-of-enum schema generation?

Did you test jmxtrans (as an example of something we don't religiously test but which binds to addresses, maybe there are others too).

**Comment by scholzj** [comment]
> Could we add a test for the list-of-enum schema generation?

Thanks I forgot about these.

> Did you test jmxtrans (as an example of something we don't religiously test but which binds to addresses, maybe there are others too).

I did not. TBH, I have no idea how it works. But this PR does not change anything, it just gives users more options. The _default_ values without this will be still the same as after this PR. So what worked before should keep working. And if anything has issues with IPv6 or DualStack, this gives users a way how to control it.

**Review by tombentley — APPROVED** [review]

File: api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java [hunk 4]
```diff
@@ -193,6 +199,36 @@ public void setMaxConnectionCreationRate(Integer maxConnectionCreationRate) {
         this.maxConnectionCreationRate = maxConnectionCreationRate;
     }
 
+    @Description("Specifies which IP Family Policy should be used by this service. " +
+            "Available options are `SingleStack` (a single IP family), " +
+            "`PreferDualStack` (two IP families on dual-stack configured clusters or a single IP family on single-stack clusters), " +
+            "and `RequireDualStack` (two IP families on dual-stack configured clusters, otherwise fail). " +
+            "If unspecified, Kubernetes will choose the default value based on the service type. " +
+            "Available on Kubernetes 1.20 and newer.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java:207]:
```suggestion
    @Description("Specifies the IP Family Policy used by the service. " +
            "Available options are `SingleStack`, `PreferDualStack` and `RequireDualStack`. " +
            "`SingleStack` is for a single IP family. " +
            "`PreferDualStack` is for two IP families on dual-stack configured clusters or a single IP family on single-stack clusters. " +
            "`RequireDualStack` fails unless there are two IP families on dual-stack configured clusters. " +
            "If unspecified, Kubernetes will choose the default value based on the service type. " +
            "Available on Kubernetes 1.20 and newer.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java [hunk 5]
```diff
@@ -193,6 +199,36 @@ public void setMaxConnectionCreationRate(Integer maxConnectionCreationRate) {
...
+    }
+
+    public void setIpFamilyPolicy(IpFamilyPolicy ipFamilyPolicy) {
+        this.ipFamilyPolicy = ipFamilyPolicy;
+    }
+
+    @Description("Specifies which IP Families should be used by this service. " +
+            "Available options are `IPv4` and `IPv6. " +
+            "If unspecified, Kubernetes will choose the default value based on the `ipFamilyPolicy` setting. " +
+            "Available on Kubernetes 1.20 and newer.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java:221]:
```suggestion
    @Description("Specifies the IP Families used by the service. " +
            "Available options are `IPv4` and `IPv6. " +
            "If unspecified, Kubernetes will choose the default value based on the `ipFamilyPolicy` setting. " +
            "Available on Kubernetes 1.20 and newer.")
```

**Review by PaulRMellor — APPROVED** [review]

**Pull request merged**